marcdown_project/frontend/src/pages/Editor.js

```diff
@@ -1,108 +1,220 @@
+            readOnly: false,
+            isOwner: true,
+            inputSharer: "",
+            sharedWith: [],
+            redirectToMainPage: false
+        };
+
+        this.modal = React.createRef();
+    }
+
+    _saveNote() {
+        if (this.state.existsInDatabase) {
+            const patchText = dmp.patch_toText(dmp.patch_make(this.state.previousSavedText, this.state.input));
+
+            query(`/api/note/${this.state.noteId}/`, "PATCH", {
+                diff: patchText
+            }).then((result) => { });
+            this.state.previousSavedText = this.state.input;
+        } else {
+            // Create a new note
+            query(`/api/note/`, "POST", {
+                public: false,
+                readOnly: false,
+                sharedWith: [],
+                content: this.state.input
+            }).then((result) => {
+                location.hash = `/note/${result.id}`;
+                this.setState({ existsInDatabase: true, noteId: result.id, previousSavedText: this.state.input });
+            });
+        }
+    }
+
+    _setReadOnly(readOnly) {
+        query(`/api/note/${this.state.noteId}/`, "PUT", {
+            readOnly: readOnly
+        }).then((result) => { });
+        this.setState({ readOnly: readOnly });
+    }
+
+    _setPublic(isPublic) {
+        query(`/api/note/${this.state.noteId}/`, "PUT", {
+            public: isPublic
+        }).then((result) => { });
+        this.setState({ public: isPublic });
+    }
+
+    _addSharedUser() {
+        let newSharedArray = [...this.state.sharedWith, this.state.inputSharer];
+        this.setState({ inputSharer: "" });
+
+        query(`/api/note/${this.state.noteId}/`, "PUT", {
+            sharedWith: newSharedArray
+        }).then((result) => {
+            // Remove bad sharers and update
+            if (result.badSharerNames.length > 0) {
+                this.setState({ inputSharer: "User not found" });
+                newSharedArray.splice(newSharedArray.indexOf(result.badSharerNames), 1);
+            }
+
+            this.setState({ sharedWith: newSharedArray });
+        });
     }
-  }
 
-  _loadFromDatabase(id) {
-    id = parseInt(id);
+    _removeSharedUser(index) {
+        let newSharedArray = this.state.sharedWith;
+        newSharedArray.splice(index, 1);
 
-    if (isNaN(id)) {
-      this.setState({ defaultInput: "# Could not load this note\n\nInvalid ID specified\n\nEdit this note to create a new one" });
-    } else {
-      query(`/api/note/${id}`).then((result) => {
-        if (result.id === undefined) {
-          this.setState({ defaultInput: "# Could not load this note\n\n**Error detail**: " + result.detail + "\n\nMake sure you have the permission to read this note\n\nEdit this note to create a new one" });
+        query(`/api/note/${this.state.noteId}/`, "PUT", {
+            sharedWith: newSharedArray
+        }).then((result) => {
+        });
+
+        this.setState({ sharedWith: newSharedArray });
```

Comment:
Shouldn't this be done when the API confirms the new sharers ?

Reply:
This is done so the user instantly disappear when the user click. This should only fail if there is an issue with the API